consumer/transaction.go

```diff
@@ -187,80 +201,98 @@ func txnRead(s *shard, txn, prev *transaction, env EnvelopeOrError, ok bool) err
 	readHeadGauge.WithLabelValues(env.Journal.Name.String()).Set(float64(env.End))
 	// End DEPRECATED metrics.
 
-	if !s.sequencer.QueueUncommitted(env.Envelope) {
-		// Envelope is a continuation of an ongoing sequence of messages.
-		// No need to dequeue for committed messages (there are none).
-		return nil
-	}
+	// |env| is read-uncommitted. Queue and act on its sequencing outcome.
+	var outcome = s.sequencer.QueueUncommitted(env.Envelope)
 
-	for {
-		switch deq, err := s.sequencer.DequeCommitted(); err {
-		case nil:
-			if err = txnConsume(s, txn, deq); err != nil {
-				return err
-			}
+	switch outcome {
+	case message.QueueAckCommitReplay:
 
-		case io.EOF:
-			if txn.consumedCount != 0 {
-				// We're inside a current transaction. Take no action.
-			} else if txn.barrierCh != nil {
-				// We're outside of a current transaction, but a previous one is still
-				// committing. Extend it's readThrough, which hasn't yet been reported
-				// as progress, to include this no-op envelope.
-				prev.readThrough[env.Journal.Name] = env.End
-			} else {
-				// We're currently idle, and must update progress to reflect that
-				// |env| was read. If we didn't do this, then a Stat RPC
-				// could stall indefinitely if its read-through offsets include a
-				// message which doesn't cause a transaction to begin (such as a
-				// duplicate ACK), and no further messages are forthcoming.
-				signalProgress(s, func(readThrough, _ pb.Offsets) {
-					readThrough[env.Journal.Name] = env.End
-				})
-			}
-			return nil
+		var journal, from, to = s.sequencer.ReplayRange()
+		var it message.Iterator
 
-		case message.ErrMustStartReplay:
-			var from, to = s.sequencer.ReplayRange()
-			var it message.Iterator
-
-			if mp, ok := s.svc.App.(MessageProducer); ok {
-				it = mp.ReplayRange(s, s.store, deq.Journal.Name, from, to)
-			} else {
-				var rr = client.NewRetryReader(s.ctx, s.ajc, pb.ReadRequest{
-					Journal:    deq.Journal.Name,
-					Offset:     from,
-					EndOffset:  to,
-					Block:      true,
-					DoNotProxy: !s.ajc.IsNoopRouter(),
-				})
-				it = message.NewReadUncommittedIter(rr, s.svc.App.NewMessage)
-			}
-			s.sequencer.StartReplay(it)
+		if mp, ok := s.svc.App.(MessageProducer); ok {
+			it = mp.ReplayRange(s, s.store, journal, from, to)
+		} else {
+			var rr = client.NewRetryReader(s.ctx, s.ajc, pb.ReadRequest{
+				Journal:    journal,
+				Offset:     from,
+				EndOffset:  to,
+				Block:      true,
+				DoNotProxy: !s.ajc.IsNoopRouter(),
+			})
+			it = message.NewReadUncommittedIter(rr, s.svc.App.NewMessage)
+		}
+		s.sequencer.StartReplay(it)
+		fallthrough
 
-		default:
-			return errors.WithMessage(err, "dequeCommitted")
+	case message.QueueAckCommitRing, message.QueueAckEmpty, message.QueueOutsideCommit:
+
+		// Acknowledged messages are ready to dequeue.
+		// We don't expect io.EOF here (we'll minimally dequeue |env| itself).
+		if err := s.sequencer.Step(); err != nil {
```

Comment:
Just checking, do we need to call `Step` in a loop here, or does one call take care of all the messages implied by `QueueAckCommitRing`?

Reply:
No, we definitely want to step _just one_ message because consumption must be driven by the top-level `txnStep`, which is checking for a timeout or for the consumer app to defer the message or lots of other things.

Basically everything got inverted, and the tight loop that used to process committed messages instead now makes progress by single polled calls to `txnStep`, each of which handles a single dequeued message. The trade-off is that we can precisely end the transaction whenever we want (timeout, or because the app deferred the message).